### PR TITLE
[Feature] toFactory implementation

### PR DESCRIPTION
**Pull request by SieR-VR**
- [x] implement #69

### CODE REVIEW

File: src/async.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ export type ListFactoryFunc<T, K extends keyof T, U = T> = keyof T extends K
   ? (count: number, item?: RecPartial<T>) => Promise<U[]>
   : (count: number, item: RecPartial<T> & Omit<T, K>) => Promise<U[]>;
 
-function isPromise<T extends Object>(t: T | Promise<T>): t is Promise<T> {
+function isPromise<T>(t: T | Promise<T>): t is Promise<T> {
```
Comment (willryan) [inline src/async.ts:18]: I guess this change was made to keep `lift()` happy? But `isPromise()` will now compile with null or undefined as arguments, and at present that will cause this function to crash.  maybe change the function body to:

```typescript
  return !!t && typeof (t as any)["then"] === "function";
```

File: src/async.ts [hunk 2]
```diff
@@ -270,8 +270,11 @@ interface BaseBuild<T> {
 
 async function buildBase<T, K extends keyof T>(
   seqNum: number,
-  builder: Builder<T, K>
+  builder: Builder<T, K> | Promise<Builder<T, K>>
 ): Promise<BaseBuild<T>> {
+  if (isPromise(builder))
+    builder = await builder;
```
Comment (willryan) [inline src/async.ts:276]: I generally don't like reassignment like this, although Typescript does helpfully update the known type of `builder`.  The `lift()` function is meant to help with situations like this:
```
const resolvedBuilder = await lift(builder);
// now use resolvedBuilder going forward
```

File: src/pipeline.ts [hunk 3]
```diff
@@ -91,4 +91,8 @@ export class Pipeline<P extends Object = {}> implements PromiseLike<P> {
   ): Promise<TResult1 | TResult2> {
     return this.current.then(onfulfilled, onrejected);
   }
-}
+
+  toFactory(): Async.Factory<P> {
+    return Async.makeFactory(this.current as Promise<Async.Builder<P, keyof P>>);
```
Comment (willryan) [inline src/pipeline.ts:96]: I don't love that we have to cast here, but I can't find anything more elegant.  `this.current` _is_ a valid `Async.Builder<P>`, but unfortunately makeFactory's signature won't quite allow it.  Maybe make a comment here explaining why the cast is valid?  Also you can just say `Promise<Async.Builder<P>>` (the second generic ar defaults to `keyof P`)

**Review by willryan — CHANGES_REQUESTED** [review]
Really great work!  Thanks for taking this on, and sorry it took me a bit to get to it.  I did finally get approval to keep working on this project under my full ownership, though, so that's good.  I just a few small requests / suggestions.

File: src/pipeline.ts (hunk 3 above)
Comment (SieR-VR) [reply src/pipeline.ts]: this.current is Promise object, so I cannot help to use cast as Promise<Async.Builder<P, keyof P>>.
and I changed type parameter, thanks for review.

**Review by willryan — APPROVED** [review]
Looks good.  I'll push out a new version after this gets merged.

**Pull request merged**